### DISCUSSION TITLE
Allow static and instance methods with same name

### DISCUSSION THREAD

**Discussion by LostTime76** [dotnet/csharplang#6866] [Language Ideas]
I have a use case that I have been having trouble with for awhile. It makes sense that a class I am developing have both static and instance methods of the same name and type signature, maybe only differing in return type. The user of the class calls the static method when they know what they want and calls the instance method when an instance of that type is passed back instead.

How can I get past this issue today? Slightly different name, explicit interface implementation, etc.. All of these have warts and confuse the API surface.

I would not be asking for this if the syntax I am looking for was not already valid C# and possible today, because it already is indirectly.

```
public static class Test
{
    public static void Main(string[] args)
    {
        Foo foo = new Foo();
        foo.Do();
        Foo.Do();
    }
}

public class Foo
{
    public static void Do()
    {
        
    }
}

public static class Exts
{
    public static void Do(this Foo foo)
    {

    }
}
```

Why not just use that then? Unfortunately this only works if Foo is a top level class and there is no intention of derivation. Foo in my case is an abstract base class and derived classes want to have static methods with the same name and possibly different return types. In my case, I wish that extensions could be static instead. That way the correct abstract / concrete implementations would go through using the standard class model.. then a static extension function, which does not need to participate in all that could be defined.

Maybe this wouldn't be an issue if one of those 'extension everything' proposals gets off the ground... However, that may still require polluting the namespace with another random top level static class.

The ask would be to allow code like this:

```
public static class Test
{
    public static void Main(string[] args)
    {
        Foo.Do();
        Bar.Do();
        Fizz.Do();

        Foo foo = new Fizz();
        Bar bar = (Bar)foo;
        foo.Do();
        bar.Do();
    }
}

public abstract class Foo
{
    public abstract void Do();

    public static void Do()
    {

    }
}

public abstract class Bar : Foo
{
    public static void Do()
    {

    }
}

public class Fizz : Bar
{
    public override void Do()
    {

    }

    public static void Do()
    {

    }
}
```

The issue is obviously the name collisions with the static functions and the ambiguity within the instance Do method. Let's take Fizz as the example. This could be a solution.

```
public class Fizz : Bar
{
    public override void Do()
    {
        Do(); // Error, ambiguous - call Fizz instance Do, or Fizz / Bar / Foo static Do?

        base.Do(); // Calls Bar class instance Do
        this.Do(); // Calls Fizz instance Do
        Fizz.Do(); // Calls Fizz static Do
        Bar.Do();  // Calls Bar static Do
    }

    public static void Do()
    {
        Do(); // Calls Fizz static Do
    }
}
```

All of the normal processing remains in place. However, if an ambiguity is found, the user is forced to resolve it through qualification.

**Comment by iam3yal** [comment]
@LostTime76 It might worth having the actual names you're having trouble with as opposed to the generic names you're using in your examples, personally, I think that this would impact readability, it forces people to look into the implementation to understand the differences between `foo.Do()` and `Foo.Do()` especially when both returns `void`.

> How can I get past this issue today? Slightly different name, explicit interface implementation, etc.. All of these have warts and confuse the API surface.

but the same goes for your suggestion, can you be more specific with your examples? how would reusing the same name make things clear or clean?

**Comment by LostTime76** [comment]
Unfortunately, I can't give too many details on my specific implementation. In my case, the static and instance methods sharing the same name allow calling the method in different contexts. 

I would like to point out that I guess swift allows this behavior: https://forums.swift.org/t/same-name-allowed-for-both-static-and-instance-member/51729. I think the example here of isMainThread is the most succinct example that gets the point across. 

I could try and convince you of why I would like this behavior in my case and why its clearer; however, asking 100 different people, they will all say different things on how to implement something.

Since I believe static and instance members have different IL signatures anyways and there seems to exist ways to easily rid the ambiguity, where it exists, this should be low hanging fruit in theory. I guess, a nice to have? No doubt implementation would be much more difficult. One language saw fit to allow the behavior, why not this one? The language seems to be taking queues from others all the time.

**Reply by HaloFour** [reply]
Different languages have different rules.  In C# this is disallowed because you are allowed to invoke a method in scope without a qualifier, which would be ambiguous:

```cs
public class C {
    public void M() { }

    public static void M() { }

    public void InstanceMethod() {
        M(); // which does it mean?
    }
}
```

Yes, this could be disambiguated by requiring qualification (`this.M()` or `C.M()`), but it's also very simple to just give the one of the methods a different name.

**Reply by LostTime76** [reply]
So essentially by allowing a static method be called without a qualification, a static method can be invoked on an implicit instance of the type? Me calling M() implicitly is calling this.M(). I thought we weren't allowed to call static methods off of instances of objects? I could dig up the thread where somebody was asking for the ability to call static methods off of instances, but the team seemingly shut them down saying it was confusing. 

Just thought I would poke a hole in the reasoning of the language design. Why am I allowed to call a static method without qualification within the class? Looks like this.StaticMethod() to me.

I know internally there is a special case for allowing this behavior and it does not use 'this' to call the static method. However, for all the world as an onlooker, I am implicitly using 'this'. Talk about confusing.

**Reply by CyrusNajmabadi** [reply]
> However, for all the world as an onlooker, I am implicitly using 'this'.

If we were doing it again, we might perhaps not allow this.  But this was a 1.0 decision that we're stuck with.

**Reply by LostTime76** [reply]
Good enough for me. I was just making a point.

Thanks

**Reply by CyrusNajmabadi** [reply]
> So essentially by allowing a static method be called without a qualification, a static method can be invoked on an implicit instance of the type?

No.  It's simply that lookup allows for both static and instance methods to be found.  If an instance method is found, it is invoked on 'this'.  If a static method is found, it is invoked statically.  THere is no concept of a static method invoked on an implicit instance (or an explicit instance for that matter) since static methods have no 'this' parameter in the first place (that's how they're static :)).

**Reply by LostTime76** [reply]
If I was looking at the language as just a user of it, it would look like the language is calling this.StaticMethod() using the same rules everywhere else. That is why I didn't have to qualify with Type.StaticMethod(). This is not saying anything about the actual implementation, just that its a confusing rule. Inside the class I don't have to qualify it, but outside the class I do. Therefore, it 'looks' like its working just like how I don't have to qualify InstanceMethod() within the class, because its using 'this' behind the scenes (this.InstanceMethod()).

**Reply by TahirAhmadov** [reply]
The problem isn't just that C# could not have made some kind of rule to "prefer" instance vs. static or vice-versa, but that allowing it now would open the door for many potential breaking changes - for example, adding a static method with a signature matching an existing instance method would now be allowed (instead of breaking the build as it does now), and if the rule is that `Foo()` prefers the `static` method, suddenly the program's behavior changes without any warning. We have just a few years of language's existence and at least a couple of hundred programs written in it. These things have to be taken seriously.

On another note, if we were to start mandating `this.Foo()` prefixes instance calls and `TypeName.Foo()` prefixes to static calls, I would strongly prefer something like `this::Foo()` or perhaps a different syntax, for `static` methods - there are sometimes long class names with generic parameters and writing `VeryLongClassNameInComplexSoftware<string, Person, RelationshipManager<Person>>` in front of each `static` would be a huge deal breaker.

**Reply by HaloFour** [reply]
Interesting it's only an issue where the signatures are equivalent.  The compiler is perfectly fine doing overload resolution between a mix of static and instance methods with the same name.

Is this limitation also in IL or is it imposed by the C# compiler?  If the latter, how does the compiler handle an assembly where this situation might already exist?

Assuming this is an imposed limitation in C# it would seemingly be easy enough to resolve, if the member is ambiguous require qualification.  I would much rather that than to have the compiler prefer one or the other anyway.  But it's also one of those things that I doubt would bubble up to the point of being worth addressing.

**Reply by LostTime76** [reply]
Yes, in my solution in the OP, I preferred that if there was ambiguity, it would require qualification. That means, quite literally, no existing code breaks. If it compiles today, it compiles tomorrow. A dummy like myself would have to intentionally make a static &  instance method with the same name and parameters in order for the new rule of qualification  to fix ambiguity (now allowed) to take effect. 

In my OP, I actually was going to present a few options, one of which being the compiler chose one, but it was a rule. Then you would know the rule exists and know how to call the static method. For example, the compiler would always choose the instance method, then you would be required to qualify it to call the static method. Then I looked at it and said.. nah, throw an error and force the coder to intentionally fix it. 

I imagine the ambiguous handling would be similar to how it is today. When the compiler cannot resolve a method call because it doesn't know which to call, it throws an error and you have to change the code:

```
public void Method(bool flag = true) { }
public void Method() { }

Method(); // Compiler ambiguity error
```

**Reply by TahirAhmadov** [reply]
Good point about forcing qualification in ambiguous situations. I also just noticed it in the OP. 

Personally, though, I would question the need for it.

> In my case, the static and instance methods sharing the same name allow calling the method in different contexts.

This is a red flag to me. If an operation can be performed statically, why have the instance version? If the static version relies on a static instance of the class behind the scenes, why create all these "wrappers" - why not just expose the said static reference to the instance? I would question the architecture here. This actually makes me thankful for the existence of this limitation. 
_PS. Following my own logic, I think we need an analyzer to outright prohibit mixing static/instance methods with the same name regardless of the signature._

I think there was another discussion recently which requested this, correct? That one, if I recall, was about "transitioning" to `static` for simplicity/performance, and that motivation was more convincing than this discussion.

**Reply by svick** [reply]
@HaloFour 

> Is this limitation also in IL or is it imposed by the C# compiler? If the latter, how does the compiler handle an assembly where this situation might already exist?

ILAsm doesn't have any issues compiling a class containing a static and instance methods with the same signature.

When I inherit that class in C# and try to call the method I get:

> error CS0121: The call is ambiguous between the following methods or properties: 'C.M()' and 'C.M()'

**Comment by iam3yal** [comment]
@LostTime76 You can do something like this:
```csharp
var thread = new Thread();

thread.IsMainThread();
Thread.IsMainThread(thread);
Thread.IsMainThread();

public sealed class Thread
{
    public static bool IsMainThread(Thread? thread = null)
    {
        if (thread is not null)
        {
            return thread.IsMainThread();
        }

        Console.WriteLine("Do something else here");
        
        return false;
    }
    
    public bool IsMainThread()
    {
        Console.WriteLine("Do something here");
        
        return false;
    }
}
```
Now, I'd definitely go with another name for both or design it differently so I wouldn't have this problem to begin with but the above example might work for you, dunno.

**Reply by LostTime76** [reply]
Yeah, I am trying to find the best workaround. I was thinking about giving one of the methods a different type signature with the method parameters, but its a bit of a hack. I will figure something out.

Thanks
